internal/tescale/scale.go

```diff
@@ -0,0 +1,267 @@
+	p.rw.Lock()
+	defer p.rw.Unlock()
+
+	p.index++
+	return p.ips[p.index]
+}
+
+// AllIPs returns all IPs in the pool
+func (p *IPPool) AllIPs() []string {
+	return append([]string{}, p.ips...)
+}
+
+// IDPool for NH and NHG IDs
+type IDPool struct {
+	nhIndex  uint64
+	nhgIndex uint64
+	rw       sync.RWMutex
+}
+
+// NewIDPool creates a new IDPool
+func NewIDPool(base uint64) *IDPool {
+	return &IDPool{
+		nhIndex:  base,
+		nhgIndex: base,
+	}
+}
+
+// NextNHID returns the next NHID
+func (p *IDPool) NextNHID() uint64 {
+	p.rw.Lock()
+	defer p.rw.Unlock()
+
+	p.nhIndex++
+	return p.nhIndex
+}
+
+// NextNHGID returns the next NHGID
+func (p *IDPool) NextNHGID() uint64 {
+	p.rw.Lock()
+	defer p.rw.Unlock()
+
+	p.nhgIndex++
+	return p.nhgIndex
+}
+
+// VRFConfig holds NH, NHG and IPv4 entries for the VRF.
+type VRFConfig struct {
+	Name      string
+	NHs       []fluent.GRIBIEntry
+	NHGs      []fluent.GRIBIEntry
+	V4Entries []fluent.GRIBIEntry
+}
+
+// Param TE holds scale parameters.
+type Param struct {
+	V4TunnelCount         int
+	V4TunnelNHGCount      int
+	V4TunnelNHGSplitCount int
+	EgressNHGSplitCount   int
+	V4ReEncapNHGCount     int
+}
+
+// BuildVRFConfig creates scale new scale VRF configurations.
+func BuildVRFConfig(dut *ondatra.DUTDevice, egressIPs []string, param Param) []*VRFConfig {
+	v4TunnelIPAddrs := NewIPPool(iputil.GenerateIPs(V4TunnelIPBlock, param.V4TunnelCount))
+	v4VIPAddrs := NewIPPool(iputil.GenerateIPs(V4VIPIPBlock, (param.V4TunnelNHGCount*param.V4TunnelNHGSplitCount)+2))
+	v4EgressIPAddrs := NewIPPool(egressIPs)
+
+	defaultVRF := deviations.DefaultNetworkInstance(dut)
+	vrfTConf := &VRFConfig{Name: VRFT}
+	vrfRConf := &VRFConfig{Name: VRFR}
+	vrfRDConf := &VRFConfig{Name: VRFRD}
+	vrfDefault := &VRFConfig{Name: defaultVRF}
+	idPool := NewIDPool(10000)
+
+	// VRF_T:
+
+	nhgID := idPool.NextNHGID()
+	nhID := idPool.NextNHID()
+	nhgRedirectToVrfR := nhgID
+	// build backup NHG and NH.
+	vrfDefault.NHs = append(vrfDefault.NHs,
+		fluent.NextHopEntry().WithIndex(nhID).WithNetworkInstance(defaultVRF).WithNextHopNetworkInstance(VRFR),
+	)
+	vrfDefault.NHGs = append(vrfDefault.NHGs,
+		fluent.NextHopGroupEntry().WithID(nhgRedirectToVrfR).AddNextHop(nhID, 1).WithNetworkInstance(defaultVRF),
+	)
+
+	// Build IPv4 entry and related NHGs and NHs.
+	// * Mapping tunnel IP per the IP -> NHG ratio
+	// * Each NHG has unique NHs.
+	// * Each NHG has the same backup to Repair VRF.
+	tunnelNHGRatio := param.V4TunnelCount / param.V4TunnelNHGCount
+	for idx, ip := range v4TunnelIPAddrs.AllIPs() {
+		if idx%tunnelNHGRatio == 0 {
+			nhgID = idPool.NextNHGID()
+
+			// Build NHs and link NHs to NHG.
+			for i := 0; i < param.V4TunnelNHGSplitCount; i++ {
+				vip := v4VIPAddrs.NextIP()
+				nhID = idPool.NextNHID()
+				vrfDefault.NHs = append(vrfDefault.NHs,
+					fluent.NextHopEntry().WithIndex(nhID).WithNetworkInstance(defaultVRF).WithIPAddress(vip),
+				)
+				vrfDefault.NHGs = append(vrfDefault.NHGs,
```

Comment:
why are we appending the same NHG ID multiple times.  Shouldnt the NHs be created inside the 2nd for loop and the NHG added to "vrfDefault.NHGs" once outside with the no. of nhs = param.V4TunnelNHGSplitCount